storage/innobase/include/arch0arch.h

```diff
@@ -0,0 +1,1984 @@
+  @param[in]     buffer     buffer filled with 8 byte page ids; the format is
+  specific to SE. For InnoDB it is space_id (4 bytes) followed by page number
+  (4 bytes)
+  @param[in]     buf_len    length of buffer in bytes
+  @param[in]     num_pages  number of valid page IDs in buffer
+  @param[in,out] user_ctx   user context passed to page tracking function
+  @return Operation status.
+ */
+ typedef int (*Page_Track_Callback)(MYSQL_THD thd, const unsigned char *buffer,
+                                    size_t buf_len, int num_pages,
+                                    void *user_ctx);
+
+  /** Get archived page Ids between two given LSN values.
+  Attempt to read blocks directly from in memory buffer. If overwritten,
+  copy from archived files.
+  @param[in]    thd             thread handle
+  @param[in]      cbk_func        called repeatedly with page ID buffer
+  @param[in]      cbk_ctx         callback function context
+  @param[in,out]  start_id        fetch archived page Ids from this LSN
+  @param[in,out]  stop_id         fetch archived page Ids until this LSN
+  @param[in]      buf             buffer to fill page IDs
+  @param[in]      buf_len         buffer length in bytes
+  @return error code */
+  int get_pages(MYSQL_THD thd, Page_Track_Callback cbk_func, void *cbk_ctx,
+                lsn_t &start_id, lsn_t &stop_id, byte *buf, uint buf_len);
+
+  /** Set the latest stop LSN to the checkpoint LSN at the time it's called. */
+  void post_recovery_init();
+
+  /** Recover the archiver system at the time of startup. Recover information
+  related to all the durable groups and start archiving if any group was active
+  at the time of crash/shutdown.
+  @return error code */
+  dberr_t recover();
+
+#ifdef UNIV_DEBUG
+  /** Print information related to the archiver for debugging purposes. */
+  void print();
+#endif
+
+  /** Set the state of the archiver system to read only. */
+  void set_read_only_mode() { m_state = ARCH_STATE_READ_ONLY; }
+
+  /** Check if archiver system is in initial state
+  @return true, if page ID archiver state is #ARCH_STATE_INIT */
+  bool is_init() const { return (m_state == ARCH_STATE_INIT); }
+
+  /** Check if archiver system is active
+  @return true, if page ID archiver state is #ARCH_STATE_ACTIVE or
+  #ARCH_STATE_PREPARE_IDLE. */
+  bool is_active() const {
+    return (m_state == ARCH_STATE_ACTIVE || m_state == ARCH_STATE_PREPARE_IDLE);
+  }
+
+  /** @return true if in abort state */
+  bool is_abort() const { return (m_state == ARCH_STATE_ABORT); }
+
+  /** Get the mutex protecting concurrent start, stop operations required
+  for initialising group during recovery.
+  @return mutex */
+  mysql_mutex_t *get_mutex() { return (&m_mutex); }
+
+  /** @return operation mutex */
+  mysql_mutex_t *get_oper_mutex() { return (&m_oper_mutex); }
+
+  /** Fetch the system client context.
+  @return system client context. */
+  Page_Arch_Client_Ctx *get_sys_client() const { return (m_ctx); }
```

Comment:
The `get_mutex()` appears to be unreachable. What is the point of these? We could declare each data member `public`. If `m_ctx` really needs the pointer indirection, then it could be declared as `Page_Arch_Client_Ctx *const m_ctx;` and be `public`.